src/main/java/org/tugraz/sysds/runtime/lineage/Lineage.java

```diff
@@ -0,0 +1,172 @@
+            if (fs instanceof LocalFileSystem) {
+                Path path = new Path(fname);
+                IOUtilFunctions.deleteCrcFilesFromLocalFileSystem(fs, path);
+            }
+
+        } catch (IOException e) {
+            throw new DMLRuntimeException(e);
+        }
+    }
+
+    private static void removeInstruction(VariableCPInstruction inst) {
+        removeLineageItem(inst.getInput1().getName());
+    }
+
+    private static void moveInstruction(VariableCPInstruction inst) {
+        if (inst.getInput2().getName().equals("__pred")) {
+            removeLineageItem(inst.getInput1().getName());
+        } else {
+            LineageItem li_source = get(inst.getInput1());
+            LineageItem li;
+            ArrayList<LineageItem> lineages = new ArrayList<>();
+            if (li_source != null) {
+                lineages.add(li_source);
+                li = new LineageItem(inst.getInput2(), lineages, inst.getOpcode());
+            } else {
+                lineages.add(getOrCreate(inst.getInput1()));
+                if (inst.getInput3() != null)
+                    lineages.add(getOrCreate(inst.getInput3()));
+                li = new LineageItem(inst.getInput2(), lineages, inst.getOpcode());
+            }
+            addLineageItem(li);
+        }
+    }
+
+    private static void copyInstruction(VariableCPInstruction inst) {
+        LineageItem li_source = get(inst.getInput1());
+        LineageItem li;
+        ArrayList<LineageItem> lineages = new ArrayList<>();
+        if (li_source != null) {
+            lineages.add(li_source);
+            li = new LineageItem(inst.getInput2(), lineages, inst.getOpcode());
+        } else {
+            lineages.add(getOrCreate(inst.getInput1()));
+            li = new LineageItem(inst.getInput2(), lineages, inst.getOpcode());
+        }
+        addLineageItem(li);
+    }
+
+    private static void createVariableInstruction(VariableCPInstruction inst) {
+        ArrayList<LineageItem> lineages = new ArrayList<>();
+        lineages.add(lineage_traces.getOrDefault(inst.getInput2(),
+                new LineageItem(inst.getInput2())));
+        lineages.add(lineage_traces.getOrDefault(inst.getInput3(),
+                new LineageItem(inst.getInput3())));
+        LineageItem li = new LineageItem(inst.getInput1(), lineages, inst.getOpcode());
+        addLineageItem(li);
+    }
+
+    public static LineageItem get(CPOperand variable) {
+        if (variable == null)
+            return null;
+        return lineage_traces.get(variable.getName());
+    }
+
+    private static void removeAncestorLink(LineageItem li) {
+        if (li.getChildren().isEmpty()) {
+            ArrayList<LineageItem> ancestors = new ArrayList<>(li.getAncestors());
+            for (LineageItem an : ancestors) {
+                li.getAncestors().remove(an);
+                an.getChildren().remove(li);
+                removeAncestorLink(an);
```

Comment:
Instead of copying the list (to avoid concurrent modifications), you can simply set li.ancestors to an empty list and use the old one to do the parent-child cleanups and recursive invocation.

Reply:
In my current design it is not possible to set ancestors to a new empty list because it is final.
However, I could remove the final keyword there.

Reply:
To finish my work for today, I removed the final keywords for inputs and outputs and implemented your logic as mentioned.